segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -161,14 +163,32 @@ public void setupAppend(SetupAppend setupAppend) {
                 .whenComplete((attributes, u) -> {
                     try {
                         if (u != null) {
+                            if (u instanceof StreamSegmentSealedException) {
+                                throw u;
+                            }
                             handleException(writer, setupAppend.getRequestId(), newSegment, "setting up append", u);
                         } else {
+                            SegmentProperties sp = this.store.getStreamSegmentInfo(newSegment, TIMEOUT).join();
+                            if (sp.isSealed()) {
+                                throw new StreamSegmentSealedException(newSegment);
+                            }
                             long eventNumber = attributes.getOrDefault(writer, Attributes.NULL_ATTRIBUTE_VALUE);
                             this.writerStates.putIfAbsent(Pair.of(newSegment, writer), new WriterState(eventNumber));
                             connection.send(new AppendSetup(setupAppend.getRequestId(), newSegment, writer, eventNumber));
                         }
                     } catch (Throwable e) {
-                        handleException(writer, setupAppend.getRequestId(), newSegment, "handling setupAppend result", e);
+                        if (e instanceof StreamSegmentSealedException || e.getCause() instanceof StreamSegmentSealedException) {
+                            log.info("Segment '{}' is sealed and {} cannot perform append, disabling caching of attribute value...", newSegment, writer);
+                            Map<UUID, Long> newAttributes = store.getAttributes(newSegment, Collections.singleton(writer), false, TIMEOUT).join();
+                            long eventNumber = Attributes.NULL_ATTRIBUTE_VALUE;
```

Comment:
This code looks copy pasted from above. If you use my suggestion above, you will not need this duplication.

Reply:
Yes Andrei.